domain/machine/service/provider.go

```diff
@@ -53,98 +54,92 @@ func NewProviderService(
 	defer span.End()
 
-	// Machine names must not contain container child names, so they must be
-	// singular machine name.
-	if machineName.IsContainer() {
-		return "", errors.Errorf("machine name %q cannot be a container", machineName)
-	} else if parentName.IsContainer() {
-		return "", errors.Errorf("parent machine name %q cannot be a container", parentName)
-	}
-
-	// The machine name then becomes, the <parent name>/scope/<machine name>.
-	// TODO (stickupkid): Use the placement directive to determine the
-	// the scope.
-	name, err := parentName.NamedChild("lxd", machineName.String())
-	if err != nil {
-		return "", errors.Errorf("creating machine name from parent %q and machine %q: %w", parentName, machineName, err)
-	}
-	if err := name.Validate(); err != nil {
-		return "", errors.Errorf("validating machine name %q: %w", name, err)
-	}
-
 	provider, err := s.providerGetter(ctx)
 	if err != nil {
-		return "", errors.Errorf("getting provider for machine %q: %w", name, err)
+		return "", "", errors.Errorf("getting provider for create machine with parent %q: %w", parentName, err)
 	}
 	if err := provider.PrecheckInstance(ctx, environs.PrecheckInstanceParams{}); err != nil {
-		return "", errors.Errorf("prechecking instance for machine %q: %w", name, err)
+		return "", "", errors.Errorf("prechecking instance for create machine with parent %q: %w", parentName, err)
 	}
 
-	// Make new UUIDs for the net-node and the machine.
+	// Make new UUIDs for the machine.
 	// We want to do this in the service layer so that if retries are invoked at
 	// the state layer we don't keep regenerating.
-	nodeUUID, machineUUID, err := createUUIDs()
+	machineUUID, err := machine.NewUUID()
 	if err != nil {
-		return "", errors.Errorf("creating machine %q: %w", name, err)
+		return "", "", errors.Capture(err)
 	}
-	err = s.st.CreateMachineWithParent(ctx, name, nodeUUID, machineUUID)
+	parentUUID, err := s.st.GetMachineUUID(ctx, parentName)
 	if err != nil {
-		return machineUUID, errors.Errorf("creating machine %q: %w", name, err)
+		return "", "", errors.Errorf("getting parent UUID for machine %q: %w", parentName, err)
+	}
+	stateArgs, err := s.makeCreateMachineArgs(args, machineUUID)
+	if err != nil {
+		return "", "", errors.Capture(err)
+	}
+	machineName, err := s.st.CreateMachineWithParent(ctx, stateArgs, parentUUID.String())
+	if err != nil {
+		return "", "", errors.Errorf("creating machine with parent %q: %w", parentName, err)
 	}
 
-	if err := recordCreateMachineStatusHistory(ctx, s.statusHistory, name, s.clock); err != nil {
+	if err := recordCreateMachineStatusHistory(ctx, s.statusHistory, machineName, s.clock); err != nil {
 		s.logger.Infof(ctx, "failed recording machine status history: %w", err)
 	}
 
-	return machineUUID, nil
+	return machineUUID, machineName, nil
+}
+
+func (s *ProviderService) makeCreateMachineArgs(args CreateMachineArgs, machineUUID machine.UUID) (domainmachine.CreateMachineArgs, error) {
```

Comment:
I'm guessing we'll need to validate all of these arguments?

Reply:
Yes, follow-up PR.